### ISSUE TITLE
Live Reload stopped working since 3.2.0.CR1

### ISSUE THREAD

**Issue by Eng-Fouad** [quarkusio/quarkus#34642]
### Describe the bug

Here is the full stacktrace:

```
2023-07-10 11:48:15,947 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (vert.x-worker-thread-1) Failed to start quarkus [Error Occurred After Shutdown]: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.deployment.steps.MainClassBuildStep#build threw an exception: java.lang.RuntimeException: Unable to serialize objects of type class java.util.ImmutableCollections$ListN to bytecode as it has no default constructor
	at io.quarkus.deployment.recording.BytecodeRecorderImpl$46.createValue(BytecodeRecorderImpl.java:1534)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl$DeferredArrayStoreParameter$1.write(BytecodeRecorderImpl.java:2021)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl$FixedMethodContext.writeInstruction(BytecodeRecorderImpl.java:2140)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl$SplitMethodContext.writeInstruction(BytecodeRecorderImpl.java:2097)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl$DeferredArrayStoreParameter.doPrepare(BytecodeRecorderImpl.java:2018)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl$DeferredParameter.prepare(BytecodeRecorderImpl.java:1971)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl$47.doPrepare(BytecodeRecorderImpl.java:1554)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl$DeferredParameter.prepare(BytecodeRecorderImpl.java:1971)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.writeBytecode(BytecodeRecorderImpl.java:554)
	at io.quarkus.deployment.steps.MainClassBuildStep.writeRecordedBytecode(MainClassBuildStep.java:479)
	at io.quarkus.deployment.steps.MainClassBuildStep.build(MainClassBuildStep.java:261)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:909)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:282)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at java.base/java.lang.Thread.run(Thread.java:1623)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)

	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:336)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.reloadExistingApplication(AugmentActionImpl.java:266)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.reloadExistingApplication(AugmentActionImpl.java:60)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.restartApp(IsolatedDevModeMain.java:196)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.restartCallback(IsolatedDevModeMain.java:179)
	at io.quarkus.deployment.dev.RuntimeUpdatesProcessor.doScan(RuntimeUpdatesProcessor.java:541)
	at io.quarkus.deployment.dev.RuntimeUpdatesProcessor.doScan(RuntimeUpdatesProcessor.java:441)
	at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup$4.handle(VertxHttpHotReplacementSetup.java:152)
	at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup$4.handle(VertxHttpHotReplacementSetup.java:139)
	at io.vertx.core.impl.ContextBase.lambda$null$0(ContextBase.java:137)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:264)
	at io.vertx.core.impl.ContextBase.lambda$executeBlocking$1(ContextBase.java:135)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1623)
Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.deployment.steps.MainClassBuildStep#build threw an exception: java.lang.RuntimeException: Unable to serialize objects of type class java.util.ImmutableCollections$ListN to bytecode as it has no default constructor
	at io.quarkus.deployment.recording.BytecodeRecorderImpl$46.createValue(BytecodeRecorderImpl.java:1534)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl$DeferredArrayStoreParameter$1.write(BytecodeRecorderImpl.java:2021)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl$FixedMethodContext.writeInstruction(BytecodeRecorderImpl.java:2140)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl$SplitMethodContext.writeInstruction(BytecodeRecorderImpl.java:2097)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl$DeferredArrayStoreParameter.doPrepare(BytecodeRecorderImpl.java:2018)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl$DeferredParameter.prepare(BytecodeRecorderImpl.java:1971)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl$47.doPrepare(BytecodeRecorderImpl.java:1554)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl$DeferredParameter.prepare(BytecodeRecorderImpl.java:1971)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.writeBytecode(BytecodeRecorderImpl.java:554)
	at io.quarkus.deployment.steps.MainClassBuildStep.writeRecordedBytecode(MainClassBuildStep.java:479)
	at io.quarkus.deployment.steps.MainClassBuildStep.build(MainClassBuildStep.java:261)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:909)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:282)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at java.base/java.lang.Thread.run(Thread.java:1623)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)

	at io.quarkus.builder.Execution.run(Execution.java:123)
	at io.quarkus.builder.BuildExecutionBuilder.execute(BuildExecutionBuilder.java:79)
	at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:160)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:332)
	... 18 more
Caused by: java.lang.RuntimeException: Unable to serialize objects of type class java.util.ImmutableCollections$ListN to bytecode as it has no default constructor
	at io.quarkus.deployment.recording.BytecodeRecorderImpl$46.createValue(BytecodeRecorderImpl.java:1534)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl$DeferredArrayStoreParameter$1.write(BytecodeRecorderImpl.java:2021)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl$FixedMethodContext.writeInstruction(BytecodeRecorderImpl.java:2140)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl$SplitMethodContext.writeInstruction(BytecodeRecorderImpl.java:2097)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl$DeferredArrayStoreParameter.doPrepare(BytecodeRecorderImpl.java:2018)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl$DeferredParameter.prepare(BytecodeRecorderImpl.java:1971)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl$47.doPrepare(BytecodeRecorderImpl.java:1554)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl$DeferredParameter.prepare(BytecodeRecorderImpl.java:1971)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.writeBytecode(BytecodeRecorderImpl.java:554)
	at io.quarkus.deployment.steps.MainClassBuildStep.writeRecordedBytecode(MainClassBuildStep.java:479)
	at io.quarkus.deployment.steps.MainClassBuildStep.build(MainClassBuildStep.java:261)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:909)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:282)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at java.base/java.lang.Thread.run(Thread.java:1623)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)

2023-07-10 11:48:15,958 INFO  [io.qua.dep.dev.RuntimeUpdatesProcessor] (vert.x-worker-thread-1) Live reload total time: 4.142s 

```

### Output of `java -version`

20

### Quarkus version or git rev

3.2.0.CR1 and 3.2.0.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

gradle 8.2

**Comment by Eng-Fouad** [comment]
Using debugger, I put breaking point here:

https://github.com/quarkusio/quarkus/blob/7c345006d63fbd31df95da04fcfd8772e43a461e/core/deployment/src/main/java/io/quarkus/deployment/recording/BytecodeRecorderImpl.java#L1517-L1537C26

I got this:

```
expectedType = {Class@20938} "class java.lang.Object"
finalCtorHandles = null
finalNonDefaultConstructorHolder = null
method = {MethodCreatorImpl@21083} "MethodCreatorImpl [declaringClassName=io/quarkus/deployment/steps/KeycloakDevConsoleProcessor$produceProviderComponent2117206243, methodDescriptor=MethodDescriptor{name='deploy_0', returnType='V', parameterTypes=[Lio/quarkus/runtime/StartupContext;, [Ljava/lang/Object;]}]"
param = {ImmutableCollections$ListN@21081}  size = 0
this = {BytecodeRecorderImpl$46@21080} 
```

**Comment by mkouba** [comment]
There seems to be a problem in the `KeycloakDevConsoleProcessor.produceProviderComponent()` build step which passes  an instance of `java.util.ImmutableCollections.ListN` as an argument of a recorder method. It would work if the expected type (i.e. type of a recorder method param) would be `java.util.List` (in that case `ArrayList` would be used to serialize/deserialize the list) but for some reason it's `java.lang.Object`.

I'm not quite sure which part of the Quarkus introduced this regression though.

@michalvavrik Could it be this commit? https://github.com/quarkusio/quarkus/commit/5b1aceabc16a119961f40276a44c80d83dbfc8d8

**Comment by michalvavrik** [comment]
Yeah, it was 5b1acea, I made it `Object` as from JSON point of view it was irrelevant (when serializing it checks actual type) and `io.quarkus.oidc.deployment.devservices.keycloak.KeycloakDevServicesConfigBuildItem#properties` returns `Object` (so I didn't have to cast it). Also at `io.quarkus.oidc.deployment.devservices.keycloak.KeycloakDevServicesConfigBuildItem#properties`,  `io.quarkus.oidc.deployment.devservices.keycloak.KeycloakDevServicesProcessor#KEYCLOAK_REALMS` is always `List`, in `io.quarkus.oidc.deployment.devservices.keycloak.KeycloakDevServicesConfigBuildItem#config` it is `String`. I was afraid that someone will get an idea to refactor it while there are no DEV UI OIDC tests yet (will add them when time allows), so I badly expected that keeping in Object was safer. I tested it with quickstarts and setting realm, fix @mkouba suggested works https://github.com/quarkusio/quarkus/pull/34649. @Eng-Fouad thank you for pointing straight to the problem.
